feat(curriculum): add password generator project to Python curriculum (#52444)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-regular-expressions-by-building-a-password-generator/656ef89b4b486b7a16077864.md

---
id: 656ef89b4b486b7a16077864
title: Step 26
challengeType: 20
dashedName: step-26
---

# --description--

> Your `constraints` list is going to store tuples in which the first item is the constraint parameter — namely `nums`, `special_chars`, `uppercase`, and `lowercase`.
> 
> Add a tuple to your list. Use `nums` as the first item and an empty string as the second item.

Reviewer: Rephrased the instruction as it seems all tuples need to be added in this step.

```suggestion
Your `constraints` list is going to store tuples in which the first item is the constraint parameter  named `nums`. 
The tuples `special_chars`, `uppercase`, and `lowercase` would be added later on.

For now, add a tuple to your list. Use `nums` as the first item and an empty string as the second item.
```

---
id: 656ef89b4b486b7a16077864
title: Step 26
challengeType: 20
dashedName: step-26
---

# --description--

> A tuple is another built-in data structure in Python. Tuples are very much like lists, but they are identified by parentheses `()`, instead of square brackets. Also, tuples are immutable, unlike lists.

Reviewer 2: ```suggestion
A tuple is another built-in data structure in Python. Tuples are very much like lists, but they are defined with parentheses `()`, instead of square brackets. Also, tuples are immutable, unlike lists.
```

```py
my_tuple = ('larch', 1, True)
```

> Your `constraints` list is going to store tuples in which the first item is a constraint parameter.

Reviewer 2: ```suggestion
Your `constraints` list is going to store tuples. The first item of each tuple will be a constraint parameter.
```

Add a tuple to your list. Use `nums` as the first item and an empty string as the second item.

# --hints--

You should add a tuple to the `constraints` list.

```js
({ test: () => assert.match(code, /constraints\s*=\s*\[\s*\(.*\)\s*\]/) })
```

You should add `(nums, '')` to the `constraints` list.

```js
({ test: () => assert.match(code, /constraints\s*=\s*\[\s*\(\s*nums\s*,\s*("|')\1\s*\)\s*\]/) })
```

# --seed--

## --seed-contents--

```py
import secrets
import string

def generate_password(length, nums, special_chars, uppercase, lowercase):
    
    # Define the possible characters for the password
    letters = string.ascii_letters
    digits = string.digits
    symbols = string.punctuation

    # Combine all characters
    all_characters = letters + digits + symbols

    while True:
        password = ''
        # Generate password
        for _ in range(length):
            password += secrets.choice(all_characters)
--fcc-editable-region--        
        constraints = []
        
--fcc-editable-region--
    return password

> #new_password = generate_password(8)
> #print(new_password)

Reviewer 2: ```suggestion
# new_password = generate_password(8)
# print(new_password)
```
